src/analysis.rs

```diff
@@ -27,358 +23,832 @@ pub fn generate_blackman(size: usize) -> impl Iterator<Item = f32> {
-    index: u8,
-    previous_cycle_index: u8,
+pub(crate) const DEFAULT_SMOOTHING_TIME_CONSTANT: f64 = 0.8;
+pub(crate) const DEFAULT_MIN_DECIBELS: f64 = -100.;
+pub(crate) const DEFAULT_MAX_DECIBELS: f64 = -30.;
+pub(crate) const DEFAULT_FFT_SIZE: usize = 2048;
+
+const MIN_FFT_SIZE: usize = 32;
+const MAX_FFT_SIZE: usize = 32768;
+
+// [spec] This MUST be a power of two in the range 32 to 32768, otherwise an
+// IndexSizeError exception MUST be thrown.
+fn assert_valid_fft_size(fft_size: usize) {
+    if !fft_size.is_power_of_two() {
+        panic!(
+            "IndexSizeError - Invalid fft size: {:?} is not a power of two",
+            fft_size
+        );
+    }
+
+    if !(MIN_FFT_SIZE..=MAX_FFT_SIZE).contains(&fft_size) {
+        panic!(
+            "IndexSizeError - Invalid fft size: {:?} is outside range [{:?}, {:?}]",
+            fft_size, MIN_FFT_SIZE, MAX_FFT_SIZE
+        );
+    }
 }
 
-impl TimeAnalyser {
-    /// Create a new TimeAnalyser
-    fn new() -> Self {
-        Self {
-            buffer: Vec::with_capacity(MAX_QUANTA),
-            index: 0,
-            previous_cycle_index: 0,
-        }
+// [spec] If the value of this attribute is set to a value less than 0 or more
+// than 1, an IndexSizeError exception MUST be thrown.
+fn assert_valid_smoothing_time_constant(smoothing_time_constant: f64) {
+    if !(0. ..=1.).contains(&smoothing_time_constant) {
+        panic!(
+            "IndexSizeError - Invalid smoothing time constant: {:?} is outside range [0, 1]",
+            smoothing_time_constant
+        );
     }
+}
 
-    /// Add samples to the ring buffer
-    fn add_data(&mut self, data: AudioRenderQuantumChannel) {
-        if self.buffer.len() < 256 {
-            self.buffer.push(data);
-        } else {
-            self.buffer[self.index as usize] = data;
+// [spec] If the value of this attribute is set to a value more than or equal
+// to maxDecibels, an IndexSizeError exception MUST be thrown.
+fn assert_valid_min_decibels(min_decibels: f64, max_decibels: f64) {
+    if min_decibels >= max_decibels {
+        panic!(
+            "IndexSizeError - Invalid min decibels: {:?} is greater than or equals to max decibels {:?}",
+            min_decibels, max_decibels
+        );
+    }
+}
+
+// [spec] If the value of this attribute is set to a value less than or equal to
+// minDecibels, an IndexSizeError exception MUST be thrown.
+fn assert_valid_max_decibels(max_decibels: f64, min_decibels: f64) {
+    if max_decibels <= min_decibels {
+        panic!(
+            "IndexSizeError - Invalid max decibels: {:?} is lower than or equals to min decibels {:?}",
+            max_decibels, min_decibels
+        );
+    }
+}
+
+// as the queue is composed of AtomicF32 having only 1 render quantum of room should be enough
+const RING_BUFFER_SIZE: usize = MAX_FFT_SIZE + RENDER_QUANTUM_SIZE;
+
+// single producer / single consumer ring buffer
+pub(crate) struct AnalyserRingBuffer {
+    buffer: Arc<Vec<AtomicF32>>,
```

Comment:
This `Arc` is not necessary

Reply:
alternatively, make it like this:
```
#[derive(Clone)]
pub(crate) struct AnalyserRingBuffer {
    buffer: Arc<[AtomicF32]>,
    write_index: Arc<AtomicUsize>,
}
```
Then you never need to wrap the thing inside an Arc, just clone them. I think it saves a level of indirection so it is more performant

Reply:
You can create an `Arc<[..]>` with `Arc::from(&vec_of_atomics[..]).` (you need an intermediate allocation that hopefully the compiler will skip)

Reply:
ok I did the alternative but kept the `Arc<Vec<AtomicF32>>`. I didn't manage to create the `array` from the `Vec`, with your snippet the build failed because it gives `Arc<&[AtomicF32]>` rather than `Arc<[AtomicF32]>.

Reply:
It was easier than I thought: 1c29c2f8ca820ab46442222935b6596c9bd635ef
Showing a nice bit of extra performance, before:

```
bench_analyser_node
  Instructions:            41181322 (-18.52477%)
  L1 Accesses:             58665001 (-17.67864%)
  L2 Accesses:               154232 (-63.80796%)
  RAM Accesses:               84974 (-1.944403%)
  Estimated Cycles:        62410251 (-18.34027%)
```

after:
```
bench_analyser_node
  Instructions:            39385732 (-22.07732%)
  L1 Accesses:             56186630 (-21.15673%)
  L2 Accesses:               158969 (-62.67475%)
  RAM Accesses:               84981 (-1.936325%)
  Estimated Cycles:        59955810 (-21.55078%)
```